src/Objects/ParticleContainer.cpp

```diff
@@ -0,0 +1,118 @@
+//
+// Created by Celik Yucel on 23.10.24.
+//
+
+    #include "ParticleContainer.h"
+    #include "spdlog/spdlog.h"
+    #include <algorithm>
+    #include <exception>
+
+    std::vector<Particle>::iterator ParticleContainer::begin() {
+        return particles.begin();
+    }
+    std::vector<Particle>::iterator ParticleContainer::end() {
+        return particles.end();
+    }
+
+    std::vector<std::pair<std::reference_wrapper<Particle>, std::reference_wrapper<Particle>>>::iterator ParticleContainer::begin_pairs() {
+        return particlePairs.begin();
+    }
+
+    std::vector<std::pair<std::reference_wrapper<Particle>, std::reference_wrapper<Particle> > >::iterator ParticleContainer::end_pairs() {
+        return particlePairs.end();
+    }
+
+
+
+    void ParticleContainer::addParticle(const Particle &particle) {
+        SPDLOG_TRACE("adding particle to container");
+        particles.push_back(particle);
+    }
+
+    void ParticleContainer::removeParticle(const Particle &particle) {
+        SPDLOG_TRACE("removing particle from container");
+        particles.erase(
+            std::remove_if(particles.begin(), particles.end(), [&particle](const Particle &p) {
+                    if (p.getID() == particle.getID()) {
+                        return true;
+                    }
+                    return false;
+            }),
+            particles.end());
+        reinitializePairs();
+    }
+
+    Particle& ParticleContainer::getParticle(int id) {
+        if ((int)particles.size() < id + 1) { //linear search
+            for (auto &p:particles) {
+                if (p.getID() == id){
+                    return p;
+                }
+            }
+        } else { //jump to pos id and compare, move to left if ids don't match (particle with smaller id has been deleted)
+            int i = id;
+            while (particles.at(i).getID() != id) {
+                i--;
+            }
+            return particles.at(i);
+        }
```

Comment:
This method operates under the assumption that few particles have been deleted, that the particle is indeed contained in the ParticleContainer and that the particles were added to the container in order of ascending ids, ideally starting with the id 0. Let us look at a Particle with id x. If no particles with an id smaller than x have been deleted and the ids start with 0, the Particle with id x is located at position x. If a few particles with an id < x have been deleted, we only need to move forward a few steps from position x in order to find the correct particle. Therefore, we can improve performance under certain circumstances, which are usually given during normal execution. The algorithm described here is performed in the **else-clause**.

If too many particles have been deleted (id+1 > particles.size()), the **if-clause** is taken to prevent accessing out-of-bounds when trying to access the particle at position id because too many particles have been deleted.

In the **else-clause**, no out-of-bounds memory access can occur if the Particle is really contained in the ParticleContainer. Even if the ids do not start with 0, as long as the particles are added in order of ascending id, the Particle with id x is guaranteed to be at a position <= x. This is due to the fact that new particles always get added to the Container via pushback, which adds them at the very last position of the vector, and the fact that the smallest possible id for a particle to have is 0.

While the method is not inherently wrong and does what it should when used correctly (we could find no occurrence of a memory error using an address sanitizer, neither locally nor in our ci pipeline) and can improve performance under the right circumstances, I agree that the method is convoluted and confusing. I will be heeding your advice and following your suggestions to replace this method to protect from errors caused by incorrect usage and improve code clarity. We are currently only using this method for tests, therefore performance is not an issue.

Reply:
> This method operates under the assumption that few particles have been deleted, that the particle is indeed contained in the ParticleContainer and that the particles were added to the container in order of ascending ids, ideally starting with the id 0. Let us look at a Particle with id x. If no particles with an id smaller than x have been deleted and the ids start with 0, the Particle with id x is located at position x. 

Makes sense.

> If a few particles with an id < x have been deleted, we only need to move forward a few steps from position x in order to find the correct particle. 
> Therefore, we can improve performance under certain circumstances, which are usually given during normal execution. 

I guess the idea is that it is faster to start iteration from the back fo the vector backwards than from the start of the vector forwards. I don't see this as faster (in general - of course, if the particle you are looking for is closer to the back, then this is better - but you don't know this).


> The algorithm described here is performed in the **else-clause**.
> 
> If too many particles have been deleted (id+1 > particles.size()), the **if-clause** is taken to prevent accessing out-of-bounds when trying to access the particle at position id because too many particles have been deleted.

Correct.

> 
> In the **else-clause**, no out-of-bounds memory access can occur if the Particle is really contained in the ParticleContainer. Even if the ids do not start with 0, as long as the particles are added in order of ascending id, the Particle with id x is guaranteed to be at a position <= x. This is due to the fact that new particles always get added to the Container via pushback, which adds them at the very last position of the vector, and the fact that the smallest possible id for a particle to have is 0.

No this is not correct. Suppose you add 10 particles to the container with ascending order. Then you delete particles 4-6. Then you try to look for particle with ID 9.

So you check `particles.at(9).getID() != id`.

If you have already resized `particles`, such that it is size 7 (no empty spaces), then this is out-of-bounds.

I guess the case where I do see some advantage is if you are specifically only/mostly deleting particles with lower IDs and not resizing the vector. But here, you will just be growing the vector larger and larger, which also is not suitable or performant. 

Better solutions to this problem involve not growing the vector more than needed. You can do swap deletes (swap to-be-deleted particle with end particles, then pop that off the vector) or you can just periodically resort the vector to remove empty spaces in the vector. 

> 
> While the method is not inherently wrong and does what it should when used correctly (we could find no occurrence of a memory error using an address sanitizer, neither locally nor in our ci pipeline) and can improve performance under the right circumstances, I agree that the method is convoluted and confusing. I will be heeding your advice and following your suggestions to replace this method to protect from errors caused by incorrect usage and improve code clarity. We are currently only using this method for tests, therefore performance is not an issue.

> we could find no occurrence of a memory error using an address sanitizer, neither locally nor in our ci pipeline

Do you mean when running the tests or just the main program?

Reply:
> No this is not correct. Suppose you add 10 particles to the container with ascending order. Then you delete particles 4-6. Then you try to look for particle with ID 9.
> 
> So you check particles.at(9).getID() != id.
> 
> If you have already resized particles, such that it is size 7 (no empty spaces), then this is out-of-bounds.

If we are looking for a particle with id 9 and size is 7, then `(int)particles.size() < id + 1` evaluates to true, so we are taking the **if-clause instead of the else-clause**. We are performing linear search and therefore not accessing out-of-bounds. This if-clause is there specifically to prevent access out-of-bounds at the back of the vector in the else-clause. Access out-of-bounds at the front of the vector is prevented by the fact that the particle will definitely be found within the first x entries if it is present within the vector, as I described earlier.
To prove that your example does not cause an error, I have written a new test case for it, which I can push to the Assignment 2 branch if you wish to try it yourself.
```
TEST(ParticleContainerTest, testGetParticleExample) {
    ParticleContainer testContainer;
   
    for (int i= 0; i < 4; i++) {
        Particle p(0); //0 is the type, not the id
        testContainer.addParticle(p);
}
    Particle p4(0); //id = 4
    testContainer.addParticle(p4);
    Particle p5(0); //id = 5
    testContainer.addParticle(p5);
    Particle p6(0); //id = 6
    testContainer.addParticle(p6);

    for (int i= 7; i < 10; i++) {
        Particle p(0); //0 is the type, not the id
        testContainer.addParticle(p);
    }
    EXPECT_EQ(testContainer.getParticles().size(), 10);

    for (Particle p : testContainer) {
        std::cout << p.getID() << ", ";
    }
    std::cout << std::endl;

    testContainer.removeParticle(p4);
    testContainer.removeParticle(p5);
    testContainer.removeParticle(p6);

    EXPECT_EQ(testContainer.getParticles().size(), 7);

    for (Particle p : testContainer) {
        std::cout << p.getID() << ", ";
    }
    std::cout << std::endl;

    EXPECT_EQ(testContainer.getParticle(9).getID(), 9);
}

```
Running this single test with an address sanitizer, I get the following output:

```
stef@laptop:~/molsim/oldVersion/MolSim/build$ ./Test --gtest_filter=ParticleContainerTest.testGetParticleExample
Note: Google Test filter = ParticleContainerTest.testGetParticleExample
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ParticleContainerTest
[ RUN      ] ParticleContainerTest.testGetParticleExample
0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 
0, 1, 2, 3, 7, 8, 9, 
[       OK ] ParticleContainerTest.testGetParticleExample (0 ms)
[----------] 1 test from ParticleContainerTest (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 1 test.

```
It is necessary to make sure that only this single test is run to guarantee that the IDs start with 0. If other test cases are run as well, it is possible that other Particles are generated first and therefore the IDs may start with a higher number, which leads to a test failure because the particle with id 9 cannot be found. During normal non-parallelized execution of a simulation, the particles always start with id 0 because only one ParticleContainer is generated. If this changes in the future, the particle id initialization and the getParticle function must be adapted accordingly.

> I guess the case where I do see some advantage is if you are specifically only/mostly deleting particles with lower IDs and not resizing the vector. But here, you will just be growing the vector larger and larger, which also is not suitable or performant.
> Better solutions to this problem involve not growing the vector more than needed. You can do swap deletes (swap to-be-deleted particle with end particles, then pop that off the vector) or you can just periodically resort the vector to remove empty spaces in the vector.

We are using std::vector::erase to remove particles from the container, so the vector does not grow more than needed. When removing a particle, this function resizes the vector and no empty spaces are left within the vector. 

>  Do you mean when running the tests or just the main program?

Both don’t produce any error with an address sanitizer, neither locally nor in the pipeline. We suspect that you may be including the test addParticleToPairs, which currently fails the address sanitizer both locally and in our ci pipeline. However, this test does not belong to Assignment 2 branch, but is instead part of the Assignment 3 branch and tests new functions added as part of Assignment 3 and therefore should not influence the grading for Assignment 2. Furthermore, the error is caused not by getParticle(), but instead by accessing deleted particles. I have temporarily removed this test, please try running the tests again or better yet, make sure to use the last commit to Assignment 2 before the deadline (the commit with the message “changed log messages from info to debug, parse -l first” to run the tests. This is also the commit that shows all 11 tests running in the CI pipeline without issue.

Reply:
I am building with gcc 11.4.0, and the commit is 09b117a7ba7f7347b1b40d2d497737d12b87c000 . Inside a build folder, I am running `cmake -DCMAKE_CXX_FLAGS="-fsanitize=address" ..` and then `make`. I then run `ctest`.

The test failures are non-deterministic. I ran `ctest` three times consecutively:

The first time, I get
```bash
      Start  1: CalculatorTest.correctXcalculations
 1/11 Test  #1: CalculatorTest.correctXcalculations ..............   Passed    0.02 sec
      Start  2: CalculatorTest.correctFIJcalculations
 2/11 Test  #2: CalculatorTest.correctFIJcalculations ............   Passed    0.01 sec
      Start  3: CalculatorTest.correctVcalculations
 3/11 Test  #3: CalculatorTest.correctVcalculations ..............   Passed    0.01 sec
      Start  4: CalculatorTest.correctXFVcalculations
 4/11 Test  #4: CalculatorTest.correctXFVcalculations ............   Passed    0.01 sec
      Start  5: CalculatorTest.correctLJcalculations
 5/11 Test  #5: CalculatorTest.correctLJcalculations .............   Passed    0.02 sec
      Start  6: CuboidTest.testGettersAndConstructor
 6/11 Test  #6: CuboidTest.testGettersAndConstructor .............   Passed    0.01 sec
      Start  7: ParticleContainerTest.StrctureAfterAddParticle
 7/11 Test  #7: ParticleContainerTest.StrctureAfterAddParticle ...   Passed    0.01 sec
      Start  8: ParticleContainerTest.PairsAfterAddParticle
 8/11 Test  #8: ParticleContainerTest.PairsAfterAddParticle ......   Passed    0.01 sec
      Start  9: ParticleGeneratorTest.checkContainerSize
 9/11 Test  #9: ParticleGeneratorTest.checkContainerSize .........   Passed    0.01 sec
      Start 10: ParticleGeneratorTest.checkParticlePositions
10/11 Test #10: ParticleGeneratorTest.checkParticlePositions .....   Passed    0.02 sec
      Start 11: ParticleGeneratorTest.checkOtherVariables
11/11 Test #11: ParticleGeneratorTest.checkOtherVariables ........***Exception: SegFault  0.10 sec

91% tests passed, 1 tests failed out of 11

Total Test time (real) =   0.33 sec

The following tests FAILED:
	 11 - ParticleGeneratorTest.checkOtherVariables (SEGFAULT)
```

The second time, I get
```bash
      Start  1: CalculatorTest.correctXcalculations
 1/11 Test  #1: CalculatorTest.correctXcalculations ..............***Exception: SegFault  0.09 sec
      Start  2: CalculatorTest.correctFIJcalculations
 2/11 Test  #2: CalculatorTest.correctFIJcalculations ............   Passed    0.01 sec
      Start  3: CalculatorTest.correctVcalculations
 3/11 Test  #3: CalculatorTest.correctVcalculations ..............***Exception: SegFault  0.09 sec
      Start  4: CalculatorTest.correctXFVcalculations
 4/11 Test  #4: CalculatorTest.correctXFVcalculations ............   Passed    0.01 sec
      Start  5: CalculatorTest.correctLJcalculations
 5/11 Test  #5: CalculatorTest.correctLJcalculations .............   Passed    0.02 sec
      Start  6: CuboidTest.testGettersAndConstructor
 6/11 Test  #6: CuboidTest.testGettersAndConstructor .............***Exception: SegFault  0.09 sec
      Start  7: ParticleContainerTest.StrctureAfterAddParticle
 7/11 Test  #7: ParticleContainerTest.StrctureAfterAddParticle ...   Passed    0.01 sec
      Start  8: ParticleContainerTest.PairsAfterAddParticle
 8/11 Test  #8: ParticleContainerTest.PairsAfterAddParticle ......***Exception: SegFault  0.09 sec
      Start  9: ParticleGeneratorTest.checkContainerSize
 9/11 Test  #9: ParticleGeneratorTest.checkContainerSize .........   Passed    0.02 sec
      Start 10: ParticleGeneratorTest.checkParticlePositions
10/11 Test #10: ParticleGeneratorTest.checkParticlePositions .....   Passed    0.01 sec
      Start 11: ParticleGeneratorTest.checkOtherVariables
11/11 Test #11: ParticleGeneratorTest.checkOtherVariables ........   Passed    0.01 sec

64% tests passed, 4 tests failed out of 11

Total Test time (real) =   0.54 sec

The following tests FAILED:
	  1 - CalculatorTest.correctXcalculations (SEGFAULT)
	  3 - CalculatorTest.correctVcalculations (SEGFAULT)
	  6 - CuboidTest.testGettersAndConstructor (SEGFAULT)
	  8 - ParticleContainerTest.PairsAfterAddParticle (SEGFAULT)
```

The third time I get

```bash
      Start  1: CalculatorTest.correctXcalculations
 1/11 Test  #1: CalculatorTest.correctXcalculations ..............   Passed    0.01 sec
      Start  2: CalculatorTest.correctFIJcalculations
 2/11 Test  #2: CalculatorTest.correctFIJcalculations ............   Passed    0.01 sec
      Start  3: CalculatorTest.correctVcalculations
 3/11 Test  #3: CalculatorTest.correctVcalculations ..............   Passed    0.02 sec
      Start  4: CalculatorTest.correctXFVcalculations
 4/11 Test  #4: CalculatorTest.correctXFVcalculations ............   Passed    0.01 sec
      Start  5: CalculatorTest.correctLJcalculations
 5/11 Test  #5: CalculatorTest.correctLJcalculations .............   Passed    0.02 sec
      Start  6: CuboidTest.testGettersAndConstructor
 6/11 Test  #6: CuboidTest.testGettersAndConstructor .............***Exception: SegFault  0.10 sec
      Start  7: ParticleContainerTest.StrctureAfterAddParticle
 7/11 Test  #7: ParticleContainerTest.StrctureAfterAddParticle ...   Passed    0.01 sec
      Start  8: ParticleContainerTest.PairsAfterAddParticle
 8/11 Test  #8: ParticleContainerTest.PairsAfterAddParticle ......   Passed    0.01 sec
      Start  9: ParticleGeneratorTest.checkContainerSize
 9/11 Test  #9: ParticleGeneratorTest.checkContainerSize .........   Passed    0.01 sec
      Start 10: ParticleGeneratorTest.checkParticlePositions
10/11 Test #10: ParticleGeneratorTest.checkParticlePositions .....   Passed    0.01 sec
      Start 11: ParticleGeneratorTest.checkOtherVariables
11/11 Test #11: ParticleGeneratorTest.checkOtherVariables ........***Exception: SegFault  0.11 sec

82% tests passed, 2 tests failed out of 11

Total Test time (real) =   0.38 sec

The following tests FAILED:
	  6 - CuboidTest.testGettersAndConstructor (SEGFAULT)
	 11 - ParticleGeneratorTest.checkOtherVariables (SEGFAULT)
```

With a verbose output, I get `AddressSanitizer:DEADLYSIGNAL`.

With GCC 14.2.1 on my laptop, I don't get any errors. I'm not sure why there is a discrepancy between the CI and my workstation.

I understand this is tricky, if this is only causing a problem on my workstation. We won't penalise you for a problem that is only producible on my local machine and not on the CI. Please avoid changing the CI in the master to make it easier for us to spot this.

You are correct regarding the `if`/`else` clauses; that was my misunderstanding!

> We are using std::vector::erase to remove particles from the container, so the vector does not grow more than needed. When removing a particle, this function resizes the vector and no empty spaces are left within the vector.

Ah okay. Yes, this is correct. But in this case, I really don't see the point of this reverse linear search.